src/statement.ts

```diff
@@ -0,0 +1,240 @@
+import {
+  type ValuesType,
+  type Parameters,
+  type BaseType,
+  getParameters,
+  bindValues,
+} from "./helpers/binding.js";
+import {
+  type AutoWaitConfig,
+  type Config,
+  type Signal,
+  normalize,
+} from "./helpers/index.js";
+import {
+  type ExtractedStatement,
+  type Result,
+  extractReadonly,
+  wrapResult,
+} from "./registry/utils.js";
+import { type ValueOf } from "./validator/query.js";
+import {
+  extractColumn,
+  queryAll,
+  queryFirst,
+  queryRaw,
+  exec,
+} from "./lowlevel.js";
+
+export { type ValuesType, type Parameters, type ValueOf, type BaseType };
+
+export class Statement<S = unknown> {
+  private readonly config: Config & Partial<AutoWaitConfig>;
```

Comment:
Really, everything should be hidden. I'm using the "typescript" private keyword, even though I'd rather use the ES private pattern of using # prefixes. I might still do that, but apparently that breaks things in hardhat? cc @asutula